### ISSUE TITLE
Using cached properties in Python libraries breaks keyword discovery

### ISSUE THREAD

**Issue by sanderploegsma** [robotframework/robotframework#4838]
I am working on a Python keyword library that uses the [`@functools.cached_property` decorator](https://docs.python.org/3/library/functools.html#functools.cached_property) to lazily initialize parts of the keyword library on first use. Here's a simplified example of the pattern I'm using:

``` python
import functools

class KeywordLibrary:
    @functools.cached_property
    def _lazy_property(self):
        # some expensive init logic here

    def first_keyword(self):
        pass

    def second_keyword(self):
        pass
```

However, it seems that this approach does not work with the way Robot discovers keywords: for some reason the property is evaluated while Robot is looking for keyword names, causing the initialization logic to run while the library is being imported...

I tried telling Robot that my property is not actually a keyword, and according to the [documentation](https://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#limiting-public-methods-becoming-keywords), there are four ways of doing this:

1. Prefix the method with `_`
2. Add the `@robot.api.deco.library` annotation to the Python class
3. Set `ROBOT_AUTO_KEYWORDS = False` as a Python class member
4. Add the `@robot.api.deco.not_keyword` annotation to the method

Of these options, only the last one caused Robot to stop evaluating my method, which seems like a bug to me.

---

I added a reproduction scenario below. The method is already prefixed with `_`, and adding `@library` or `ROBOT_AUTO_KEYWORDS = False` results in the same error.

``` python
import functools

class CachedPropertyLibrary:
    @functools.cached_property
    def _private_property_stay_away(self):
        raise RuntimeError("You should not have called me, Robot!")

    def do_nothing(self):
        pass
```

``` robot
*** Settings ***
Library    CachedPropertyLibrary

*** Tasks ***
Sample task
    Do nothing
```

Gives the following error:

```
==============================================================================
Tests                                                                       
==============================================================================
Sample task                                                           | FAIL |
No keyword with name 'Do nothing' found.
------------------------------------------------------------------------------
Tests                                                                 | FAIL |
1 test, 0 passed, 1 failed
==============================================================================
Output:  /tmp/tmpk7qn6gqw/output.xml
[ ERROR ] Error in file '...' on line 2: Getting keyword names from library 'CachedPropertyLibrary' failed: You should not have called me, Robot!
[ ERROR ] Error in file '...' on line 2: Getting keyword names from library 'CachedPropertyLibrary' failed: You should not have called me, Robot!
```

---

Note that a property that is manually cached does not run into the same issue, for example this works:

``` python
class KeywordLibrary:
    def __init__(self):
        _lazy_property_value = None

    @property
    def _lazy_property(self):
        if not self._lazy_property_value:
            self._lazy_property_value = # some expensive init logic here
        return self._lazy_property_value

    def first_keyword(self):
        pass

    def second_keyword(self):
        pass
```

**Comment by pekkaklarck** [comment]
Interesting. We have code to avoid calling properies, but apparently that doesn't work with `cached_property`. This code was enhanced in RF 6.1.1 to avoid issues when using `property` with `classmethod`. Have you tested with that version?

**Comment by sanderploegsma** [comment]
Thanks for the heads-up! We are currently using RobotFramework 6.0.2 which is where I encountered the issue. I just ran the reproduction code with RobotFramework 6.1.1 and indeed the error no longer occurs.

**Comment by pekkaklarck** [comment]
Thanks for the update. Apparently the fix for #4802 resolved this issue as well. Good that I made a generic fix and didn't just special case `classmethod`. I think it makes sense to add a separate test for `cached_property` before closing this issue. Hopefully you can upgrade to RF 6.1.1 easily.

**Comment by sanderploegsma** [comment]
> I think it makes sense to add a separate test for `cached_property` before closing this issue. 

Yes I agree. I would like to contribute this if you don't mind. However, I'm struggling to understand whether I can simply add this as an extra case to the existing `avoid_properties_when_creating_libraries.robot` since using `cached_property` is only possible for Python >= 3.8. Do you maybe have an example that I can base it on?

**Comment by sanderploegsma** [comment]
Actually, looking at the [source of `cached_property`](https://github.com/python/cpython/blob/3.11/Lib/functools.py#L965), I'm inclined to say that the existing test cases for a custom non-data descriptor cover this issue well enough. There is a multitude of other similar annotations that are also not explicitly tested, so I'm not sure adding tests for `cached_property` is really needed.

**Comment by pekkaklarck** [comment]
According to #4915 `cached_property` is still accessed, earlier changes just hide possible exceptions. I should have created a separate test for them. Anyway, the problem will be fixed for real in RF 7.0.
